File: src/teamlense_modules/web/services/location.service.ts
import { prisma } from "../../../shared/db/prisma";
import { env } from "../../../config/env";

type SqlRow = Record<string, unknown>;

export interface LocationSearchResult {
  id: string;
  label: string;
  address: string;
  latitude: number;
  longitude: number;
  provider: "google" | "openstreetmap";
}

const asString = (v: unknown): string => {
  if (typeof v === "string") return v;
  if (v instanceof Date) return v.toISOString();
  return "";
};

const asNumber = (v: unknown): number => {
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
};

const getString = (value: unknown): string | undefined => (typeof value === "string" ? value : undefined);

const getRecord = (value: unknown): Record<string, unknown> | undefined => {
  return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : undefined;
};

export interface OfficeLocation {
  id: string;
  organizationId: string;
  label: string;
  latitude: number;
  longitude: number;
  radiusMeters: number;
  createdAt: string;
}

/**
 * Haversine formula — returns distance in meters between two lat/lng points.
 */
function haversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const R = 6_371_000; // Earth radius in meters
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export class LocationService {
  private static schemaReady = false;

  static async searchOfficeAddresses(query: string): Promise<LocationSearchResult[]> {
    const trimmed = query.trim();
    if (trimmed.length < 3) return [];

    if (env.googlePlacesApiKey) {
      return this.searchGooglePlaces(trimmed);
    }

    return this.searchOpenStreetMap(trimmed);
  }

  private static async searchGooglePlaces(query: string): Promise<LocationSearchResult[]> {
    const response = await fetch("https://places.googleapis.com/v1/places:searchText", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Goog-Api-Key": env.googlePlacesApiKey,
        "X-Goog-FieldMask": "places.id,places.displayName,places.formattedAddress,places.location",
      },
      body: JSON.stringify({
        textQuery: query,
        maxResultCount: 8,
      }),
    });

    if (!response.ok) {
      throw new Error(`Google Places search failed with ${response.status}`);
    }

    const payload = getRecord(await response.json());
    const places = Array.isArray(payload?.places) ? payload.places : [];

    return places.flatMap((place): LocationSearchResult[] => {
      const placeRecord = getRecord(place);
      const location = getRecord(placeRecord?.location);
      const displayName = getRecord(placeRecord?.displayName);
      const latitude = Number(location?.latitude);
      const longitude = Number(location?.longitude);

      if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) {
        return [];
      }

      const label = getString(displayName?.text) || getString(placeRecord?.formattedAddress) || "Office";
      const address = getString(placeRecord?.formattedAddress) || label;

      return [{
        id: getString(placeRecord?.id) || `${latitude},${longitude}`,
        label,
        address,
        latitude,
        longitude,
        provider: "google",
      }];
    });
  }

  private static async searchOpenStreetMap(query: string): Promise<LocationSearchResult[]> {
    const url = new URL("https://nominatim.openstreetmap.org/search");
    url.searchParams.set("format", "jsonv2");
    url.searchParams.set("limit", "8");
    url.searchParams.set("q", query);

    const response = await fetch(url, {
      headers: {
        "Accept-Language": "en",
        "User-Agent": "TeamLens office location search",
      },
    });

    if (!response.ok) {
      throw new Error(`OpenStreetMap search failed with ${response.status}`);
    }

    const payload = await response.json();
    const rows = Array.isArray(payload) ? payload : [];

    return rows.flatMap((row): LocationSearchResult[] => {
      const record = getRecord(row);
      const latitude = Number(record?.lat);
      const longitude = Number(record?.lon);

      if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) {
        return [];
      }

      const address = getString(record?.display_name) || "Office";

      return [{
        id: getString(record?.place_id) || `${latitude},${longitude}`,
        label: getString(record?.name) || address.split(",")[0]?.trim() || "Office",
        address,
        latitude,
        longitude,
        provider: "openstreetmap",
      }];
    });
  }

  static async ensureSchema(): Promise<void> {
    if (this.schemaReady || !prisma.$executeRawUnsafe) return;

    await prisma.$executeRawUnsafe(`
      CREATE TABLE IF NOT EXISTS "office_locations" (
        "id"              TEXT      PRIMARY KEY,
        "organization_id" TEXT      NOT NULL,
        "label"           TEXT      NOT NULL DEFAULT 'Main Office',
        "latitude"        DOUBLE PRECISION NOT NULL,
        "longitude"       DOUBLE PRECISION NOT NULL,
        "radius_meters"   INTEGER  NOT NULL DEFAULT 200,
        "created_at"      TIMESTAMPTZ NOT NULL DEFAULT NOW(),
        "updated_at"      TIMESTAMPTZ NOT NULL DEFAULT NOW()
      )
    `);

    // Add location columns to work_sessions if not present
    await prisma.$executeRawUnsafe(`
      DO $$
      BEGIN
        IF NOT EXISTS (
          SELECT 1 FROM information_schema.columns
          WHERE table_name = 'work_sessions' AND column_name = 'latitude'
        ) THEN
          ALTER TABLE "work_sessions" ADD COLUMN "latitude" DOUBLE PRECISION;
        END IF;

        IF NOT EXISTS (
          SELECT 1 FROM information_schema.columns
          WHERE table_name = 'work_sessions' AND column_name = 'longitude'
        ) THEN
          ALTER TABLE "work_sessions" ADD COLUMN "longitude" DOUBLE PRECISION;
        END IF;

        IF NOT EXISTS (
          SELECT 1 FROM information_schema.columns
          WHERE table_name = 'work_sessions' AND column_name = 'location_type'
        ) THEN
          ALTER TABLE "work_sessions" ADD COLUMN "location_type" TEXT;
        END IF;
      END
      $$;
    `);

    this.schemaReady = true;
  }

  // ---------------------------------------------------------------------------
  // CRUD for office locations (Manager)
  // ---------------------------------------------------------------------------

  static async getOfficeLocations(organizationId: string): Promise<OfficeLocation[]> {
    await this.ensureSchema();
    if (!prisma.$queryRawUnsafe) return [];

    const rows = (await prisma.$queryRawUnsafe(
      `SELECT "id","organization_id","label","latitude","longitude","radius_meters","created_at"
       FROM "office_locations"
       WHERE "organization_id" = $1
       ORDER BY "created_at" ASC`,
      organizationId,
    )) as SqlRow[];

    return rows.map((r) => ({
      id: asString(r.id),
      organizationId: asString(r.organization_id),
      label: asString(r.label),
      latitude: asNumber(r.latitude),
      longitude: asNumber(r.longitude),
      radiusMeters: asNumber(r.radius_meters),
      createdAt: asString(r.created_at),
    }));
  }

  static async upsertOfficeLocation(
    organizationId: string,
    data: { label: string; latitude: number; longitude: number; radiusMeters: number },
  ): Promise<OfficeLocation> {
    await this.ensureSchema();

    // Check if one already exists for this org — update it; otherwise insert.
    const existing = await this.getOfficeLocations(organizationId);

    if (existing.length > 0) {
      const existingLocation = existing[0]!;
      await prisma.$executeRawUnsafe!(
        `UPDATE "office_locations"
         SET "label" = $1, "latitude" = $2, "longitude" = $3, "radius_meters" = $4, "updated_at" = NOW()
         WHERE "id" = $5`,
        data.label,
        data.latitude,
        data.longitude,
        data.radiusMeters,
        existingLocation.id,
      );
      return { ...existingLocation, ...data };
    }

    const id = crypto.randomUUID();
    await prisma.$executeRawUnsafe!(
      `INSERT INTO "office_locations" ("id","organization_id","label","latitude","longitude","radius_meters","created_at","updated_at")
       VALUES ($1,$2,$3,$4,$5,$6,NOW(),NOW())`,
      id,
      organizationId,
      data.label,
      data.latitude,
      data.longitude,
      data.radiusMeters,
    );

    return {
      id,
      organizationId,
      label: data.label,
      latitude: data.latitude,
      longitude: data.longitude,
      radiusMeters: data.radiusMeters,
      createdAt: new Date().toISOString(),
    };
  }

  /**
   * Re-evaluate currently active sessions in the organization against the
   * latest office locations. This helps dashboard reflect updated rules
   * immediately after manager changes office coordinates/radius.
   */
  static async reclassifyActiveSessions(organizationId: string): Promise<number> {
    await this.ensureSchema();
    if (!prisma.$queryRawUnsafe || !prisma.$executeRawUnsafe) return 0;

    const rows = (await prisma.$queryRawUnsafe(
      `SELECT ws."id", ws."latitude", ws."longitude"
       FROM "work_sessions" ws
       INNER JOIN "users" u ON u."id" = ws."user_id"
       WHERE u."organization_id" = $1
         AND ws."clock_out_at" IS NULL
         AND ws."location_type" <> 'manual'
         AND ws."latitude" IS NOT NULL
         AND ws."longitude" IS NOT NULL`,
      organizationId,
    )) as SqlRow[];

    let updated = 0;
    for (const row of rows) {
      const sessionId = asString(row.id);
      const latitude = asNumber(row.latitude);
      const longitude = asNumber(row.longitude);

      const locationType = await this.determineLocationType(organizationId, latitude, longitude);
      await prisma.$executeRawUnsafe(
        `UPDATE "work_sessions"
         SET "location_type" = $1, "updated_at" = NOW()
         WHERE "id" = $2`,
        locationType,
        sessionId,
      );
      updated += 1;
    }

    return updated;
  }

  static async deleteOfficeLocation(organizationId: string, locationId: string): Promise<boolean> {
    await this.ensureSchema();
    if (!prisma.$executeRawUnsafe) return false;

    await prisma.$executeRawUnsafe(
      `DELETE FROM "office_locations" WHERE "id" = $1 AND "organization_id" = $2`,
      locationId,
      organizationId,
    );
    return true;
  }

  // ---------------------------------------------------------------------------
  // Location comparison
  // ---------------------------------------------------------------------------

  /**
   * Determine location type (office/remote) for given coordinates against org's
   * office locations. Returns "office" if within radius of any office, else "remote".
   *
   * When locationSource is "ip", the coordinates are only city-level accurate (~1-50 km),
   * so we use a minimum radius of 50 km to avoid false "remote" classifications.
   */
  static async determineLocationType(
    organizationId: string,
    latitude: number,
    longitude: number,
    locationSource?: "gps" | "ip",
    accuracyMeters?: number,
  ): Promise<"office" | "remote"> {
    const offices = await this.getOfficeLocations(organizationId);

    // IP geolocation is only accurate to ~1-50 km, so use a generous minimum radius
    const IP_MIN_RADIUS = 50_000; // 50 km
    // Desktop/browser geolocation can drift significantly indoors. Use a safety floor
    // so nearby office users are not misclassified as remote.
    const GPS_MIN_RADIUS = 2_000; // 2 km

    for (const office of offices) {
      const dist = haversineDistance(latitude, longitude, office.latitude, office.longitude);
      const gpsAccuracyRadius =
        locationSource === "gps" && typeof accuracyMeters === "number" && Number.isFinite(accuracyMeters)
          ? Math.max(0, Math.round(accuracyMeters))
          : 0;
      const effectiveRadius = locationSource === "ip"
        ? Math.max(office.radiusMeters, IP_MIN_RADIUS)
        : locationSource === "gps"
          ? Math.max(office.radiusMeters, gpsAccuracyRadius, GPS_MIN_RADIUS)
          : Math.max(office.radiusMeters, gpsAccuracyRadius);

      console.log(`[Location] Distance from "${office.label}": ${Math.round(dist)}m, effective radius: ${effectiveRadius}m (source: ${locationSource ?? "unknown"}, accuracy: ${accuracyMeters ?? "n/a"})`);

      if (dist <= effectiveRadius) {
        return "office";
      }
    }

    return "remote";
  }

  /**
   * Compute daily rollup: "Office" | "Remote" | "Mixed" | null
   */
  static computeDailyLocationStatus(locationTypes: Array<string | null | undefined>): string | null {
    const valid = locationTypes.filter((t): t is string => !!t);
    if (valid.length === 0) return null;

    const hasOffice = valid.includes("office");
    const hasRemote = valid.includes("remote");

    if (hasOffice && hasRemote) return "Mixed";
    if (hasOffice) return "Office";
    if (hasRemote) return "Remote";
    return null;
  }
}

File: backend-go/internal/services/location_service.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/teamlens/backend-go/internal/models"
)

type LocationService struct {
	pool               *pgxpool.Pool
	googlePlacesAPIKey string
	httpClient         *http.Client
}

func NewLocationService(pool *pgxpool.Pool, googlePlacesAPIKey string) *LocationService {
	return &LocationService{
		pool:               pool,
		googlePlacesAPIKey: strings.TrimSpace(googlePlacesAPIKey),
		httpClient: &http.Client{
			Timeout: 8 * time.Second,
		},
	}
}

// Haversine distance in meters
func haversineMeters(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371000.0
	dLat := (lat2 - lat1) * math.Pi / 180
	dLon := (lon2 - lon1) * math.Pi / 180
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1*math.Pi/180)*math.Cos(lat2*math.Pi/180)*
			math.Sin(dLon/2)*math.Sin(dLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return R * c
}

func (s *LocationService) DetermineLocationType(ctx context.Context, organizationID string, latitude, longitude float64, locationSource *string, accuracyMeters *float64) *string {
	rows, err := s.pool.Query(ctx,
		`SELECT id, latitude, longitude, radius_meters
		 FROM office_locations
		 WHERE organization_id = $1`,
		organizationID,
	)
	if err != nil {
		return nil
	}
	defer rows.Close()

	for rows.Next() {
		var id string
		var lat, lng float64
		var radius int
		if err := rows.Scan(&id, &lat, &lng, &radius); err != nil {
			continue
		}
		dist := haversineMeters(latitude, longitude, lat, lng)
		if dist <= float64(radius) {
			r := "office"
			return &r
		}
	}

	r := "remote"
	return &r
}

func (s *LocationService) ComputeDailyLocationStatus(locationTypes []*string) *string {
	hasOffice := false
	hasRemote := false
	hasManual := false

	for _, lt := range locationTypes {
		if lt == nil {
			continue
		}
		switch *lt {
		case "office":
			hasOffice = true
		case "remote":
			hasRemote = true
		case "manual":
			hasManual = true
		}
	}

	if hasOffice && !hasRemote && !hasManual {
		r := "Office"
		return &r
	}
	if hasRemote && !hasOffice && !hasManual {
		r := "Remote"
		return &r
	}
	if hasManual && !hasOffice && !hasRemote {
		r := "Manual"
		return &r
	}
	if hasOffice || hasRemote || hasManual {
		r := "Mixed"
		return &r
	}
	return nil
}

func (s *LocationService) ListOfficeLocations(ctx context.Context, organizationID string) ([]models.OfficeLocation, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, organization_id, label, latitude, longitude, radius_meters, created_at::text
		 FROM office_locations
		 WHERE organization_id = $1
		 ORDER BY created_at ASC`,
		organizationID,
	)
	if err != nil {
		return nil, fmt.Errorf("query office locations: %w", err)
	}
	defer rows.Close()

	var locations []models.OfficeLocation
	for rows.Next() {
		var loc models.OfficeLocation
		if err := rows.Scan(&loc.ID, &loc.OrganizationID, &loc.Label, &loc.Latitude, &loc.Longitude, &loc.RadiusMeters, &loc.CreatedAt); err != nil {
			return nil, fmt.Errorf("scan office location: %w", err)
		}
		locations = append(locations, loc)
	}
	return locations, nil
}

func (s *LocationService) UpsertOfficeLocation(ctx context.Context, organizationID string, input *models.UpsertOfficeLocationInput) (*models.OfficeLocation, error) {
	// Check if a location with this label already exists
	var existingID string
	err := s.pool.QueryRow(ctx,
		`SELECT id FROM office_locations WHERE organization_id = $1 AND label = $2`,
		organizationID, input.Label,
	).Scan(&existingID)

	if err == nil && existingID != "" {
		// Update existing
		_, err := s.pool.Exec(ctx,
			`UPDATE office_locations
			 SET latitude = $1, longitude = $2, radius_meters = $3, updated_at = NOW()
			 WHERE id = $4`,
			input.Latitude, input.Longitude, input.RadiusMeters, existingID,
		)
		if err != nil {
			return nil, fmt.Errorf("update office location: %w", err)
		}
		return s.GetOfficeLocation(ctx, existingID)
	}

	id := RandomToken(16)
	_, err = s.pool.Exec(ctx,
		`INSERT INTO office_locations (id, organization_id, label, latitude, longitude, radius_meters, created_at, updated_at)
		 VALUES ($1, $2, $3, $4, $5, $6, NOW(), NOW())`,
		id, organizationID, input.Label, input.Latitude, input.Longitude, input.RadiusMeters,
	)
	if err != nil {
		return nil, fmt.Errorf("create office location: %w", err)
	}

	return s.GetOfficeLocation(ctx, id)
}

func (s *LocationService) DeleteOfficeLocation(ctx context.Context, organizationID, locationID string) error {
	tag, err := s.pool.Exec(ctx,
		`DELETE FROM office_locations WHERE id = $1 AND organization_id = $2`,
		locationID, organizationID,
	)
	if err != nil {
		return fmt.Errorf("delete office location: %w", err)
	}
	if tag.RowsAffected() == 0 {
		return fmt.Errorf("office location not found")
	}
	return nil
}

func (s *LocationService) GetOfficeLocation(ctx context.Context, id string) (*models.OfficeLocation, error) {
	var loc models.OfficeLocation
	err := s.pool.QueryRow(ctx,
		`SELECT id, organization_id, label, latitude, longitude, radius_meters, created_at::text
		 FROM office_locations WHERE id = $1`, id,
	).Scan(&loc.ID, &loc.OrganizationID, &loc.Label, &loc.Latitude, &loc.Longitude, &loc.RadiusMeters, &loc.CreatedAt)
	if err != nil {
		return nil, fmt.Errorf("get office location: %w", err)
	}
	return &loc, nil
}

// SearchLocations returns Google Places results when configured, falling back to saved office locations for local dev.
func (s *LocationService) SearchLocations(ctx context.Context, query string) ([]models.LocationSearchResult, error) {
	query = strings.TrimSpace(query)
	if query == "" {
		return []models.LocationSearchResult{}, nil
	}

	if s.googlePlacesAPIKey != "" {
		return s.searchGoogleLocations(ctx, query)
	}

	return s.searchSavedOfficeLocations(ctx, query)
}

func (s *LocationService) searchSavedOfficeLocations(ctx context.Context, query string) ([]models.LocationSearchResult, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, label, latitude, longitude
		 FROM office_locations
		 WHERE LOWER(label) LIKE LOWER($1)
		 LIMIT 10`,
		"%"+query+"%",
	)
	if err != nil {
		return nil, fmt.Errorf("search locations: %w", err)
	}
	defer rows.Close()

	var results []models.LocationSearchResult
	for rows.Next() {
		var r models.LocationSearchResult
		if err := rows.Scan(&r.ID, &r.Label, &r.Latitude, &r.Longitude); err != nil {
			continue
		}
		r.Address = r.Label
		r.Provider = "database"
		results = append(results, r)
	}

	return results, nil
}

type googleTextSearchRequest struct {
	TextQuery string `json:"textQuery"`
	PageSize  int    `json:"pageSize"`
}

type googleTextSearchResponse struct {
	Places []googlePlace `json:"places"`
	Error  *googleError  `json:"error,omitempty"`
}

type googlePlace struct {
	ID               string            `json:"id"`
	DisplayName      googleDisplayName `json:"displayName"`
	FormattedAddress string            `json:"formattedAddress"`
	Location         googleLocation    `json:"location"`
}

type googleDisplayName struct {
	Text string `json:"text"`
}

type googleLocation struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

type googleError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Status  string `json:"status"`
}

func (s *LocationService) searchGoogleLocations(ctx context.Context, query string) ([]models.LocationSearchResult, error) {
	payload, err := json.Marshal(googleTextSearchRequest{
		TextQuery: query,
		PageSize:  10,
	})
	if err != nil {
		return nil, fmt.Errorf("build google places request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://places.googleapis.com/v1/places:searchText", bytes.NewReader(payload))
	if err != nil {
		return nil, fmt.Errorf("create google places request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Goog-Api-Key", s.googlePlacesAPIKey)
	req.Header.Set("X-Goog-FieldMask", "places.id,places.displayName,places.formattedAddress,places.location")

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("call google places: %w", err)
	}
	defer resp.Body.Close()

	var googleResp googleTextSearchResponse
	if err := json.NewDecoder(resp.Body).Decode(&googleResp); err != nil {
		return nil, fmt.Errorf("decode google places response: %w", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if googleResp.Error != nil && googleResp.Error.Message != "" {
			return nil, fmt.Errorf("google places search failed: %s", googleResp.Error.Message)
		}
		return nil, fmt.Errorf("google places search failed: status %d", resp.StatusCode)
	}

	results := make([]models.LocationSearchResult, 0, len(googleResp.Places))
	for _, place := range googleResp.Places {
		if place.Location.Latitude == 0 && place.Location.Longitude == 0 {
			continue
		}

		label := strings.TrimSpace(place.DisplayName.Text)
		address := strings.TrimSpace(place.FormattedAddress)
		if label == "" {
			label = address
		}
		if address == "" {
			address = label
		}
		if label == "" {
			continue
		}

		results = append(results, models.LocationSearchResult{
			ID:        place.ID,
			Label:     label,
			Address:   address,
			Latitude:  place.Location.Latitude,
			Longitude: place.Location.Longitude,
			Provider:  "google",
		})
	}

	return results, nil
}

// ComputeDistance returns distance in meters between two coordinates
func (s *LocationService) ComputeDistance(lat1, lon1, lat2, lon2 float64) float64 {
	return haversineMeters(lat1, lon1, lat2, lon2)
}

// FormatGoogleType formats Google Places types to human-readable
func FormatGoogleType(types []string) string {
	if len(types) == 0 {
		return "Unknown"
	}
	typeMap := map[string]string{
		"street_address":              "Street Address",
		"route":                       "Route",
		"premise":                     "Premise",
		"establishment":               "Establishment",
		"point_of_interest":           "Point of Interest",
		"locality":                    "Locality",
		"administrative_area_level_1": "City",
		"country":                     "Country",
		"postal_code":                 "Postal Code",
		"subpremise":                  "Subpremise",
	}
	if label, ok := typeMap[types[0]]; ok {
		return label
	}
	return strings.Title(strings.ReplaceAll(types[0], "_", " "))
}
